statesync/chunks.go

```diff
@@ -0,0 +1,219 @@
+		waiters:    make(map[uint32][]chan<- bool),
+	}, nil
+}
+
+// Add adds a chunk to the queue, verifying its hash. It ignores chunks that already exist,
+// returning false.
+func (q *chunkQueue) Add(chunk *chunk) (bool, error) {
+	if chunk == nil {
+		return false, errors.New("cannot add nil chunk")
+	}
+	if chunk.Body == nil {
+		return false, errors.New("cannot add chunk with nil body")
+	}
+	q.Lock()
+	defer q.Unlock()
+	if q.snapshot == nil {
+		return false, errors.New("chunk queue is closed")
+	}
+	if chunk.Height != q.snapshot.Height {
+		return false, fmt.Errorf("invalid chunk height %v, expected %v", chunk.Height, q.snapshot.Height)
+	}
+	if chunk.Format != q.snapshot.Format {
+		return false, fmt.Errorf("invalid chunk format %v, expected %v", chunk.Format, q.snapshot.Format)
+	}
+	if chunk.Index >= uint32(len(q.snapshot.ChunkHashes)) {
+		return false, fmt.Errorf("received unexpected chunk %v", chunk.Index)
+	}
+	if chunk.Index < q.nextChunk || q.memChunks[chunk.Index] != nil || q.diskChunks[chunk.Index] != "" {
+		return false, nil
+	}
+	if !bytes.Equal(chunk.Hash(), q.snapshot.ChunkHashes[chunk.Index]) {
+		return false, fmt.Errorf("chunk %v hash mismatch, expected %x got %x", chunk.Index,
+			q.snapshot.ChunkHashes[chunk.Index], chunk.Hash())
+	}
+
+	// Save the chunk in memory if it is an upcoming one, otherwise spool to disk.
+	if len(q.memChunks) < bufferChunks && chunk.Index < q.nextChunk+bufferChunks {
+		q.memChunks[chunk.Index] = chunk
+	} else {
+		path := filepath.Join(q.dir, strconv.FormatUint(uint64(chunk.Index), 10))
+		err := ioutil.WriteFile(path, chunk.Body, 0644)
+		if err != nil {
+			return false, fmt.Errorf("failed to save chunk %v to file %v: %w", chunk.Index, path, err)
+		}
+		q.diskChunks[chunk.Index] = path
+	}
+
+	// If this is the next expected chunk, pass any ready chunks to Next() via channel, and close
+	// the channel when all chunks have been passed.
+	//
+	// We don't pass chunk contents, since they may be spooled to disk, and loading all of them
+	// here first would block other callers and use a ton of memory.
+	if chunk.Index == q.nextChunk {
+		for i := chunk.Index; q.memChunks[i] != nil || q.diskChunks[i] != ""; i++ {
+			q.chReady <- i
+			q.nextChunk++
+		}
+		if q.nextChunk >= uint32(len(q.snapshot.ChunkHashes)) {
+			close(q.chReady)
+		}
+	}
+
+	// Signal any waiters that the chunk has arrived.
+	for _, waiter := range q.waiters[chunk.Index] {
+		waiter <- true
+		close(waiter)
+	}
+	delete(q.waiters, chunk.Index)
+
+	return true, nil
+}
+
+// Close closes the chunk queue, cleaning up all temporary files.
+func (q *chunkQueue) Close() error {
+	q.Lock()
+	defer q.Unlock()
+	if q.snapshot == nil {
+		return nil
+	}
+	if q.nextChunk < uint32(len(q.snapshot.ChunkHashes)) {
+		close(q.chReady)
+	}
+	for _, waiters := range q.waiters {
+		for _, waiter := range waiters {
+			waiter <- false
+			close(waiter)
+		}
+	}
+	q.waiters = nil
+	q.snapshot = nil
+	err := os.RemoveAll(q.dir)
```

Comment:
maybe it makes sense to execute this first since it's the only thing which can return an error here

Reply:
Nah, that's precisely why we do it at the end. `Close()` is typically called via `defer`, and just logs any errors. We want to make sure channels are closed and goroutines are terminated even if removing the files fail.